### DISCUSSION TITLE
Template file for using CYD with Squareline Studio/VSC/Platformio

### DISCUSSION THREAD

**Discussion by peteDDD** [witnessmenow/ESP32-Cheap-Yellow-Display#102] [General]
I have worked out some of the finer points of developing projects for the CYD using Squareline Studio for UI development with Visual Studio Code/Platformio.   I am happy to share my this template which includes an interactive UI demo that also use the RGB LED.  One can then simply remove the files from the UI folder and export new files from Squareline Studio.  There are also notes and some .h files you can copy to build your own template (if for some reason you would want to).  The Squareline Studio project file is also included.

Feel free to add this to the git hub for ESP32-Cheap-Yellow-Display

This is NOT intended to be a tutorial on Squareline Studio.  For that I can suggest the following:
1: the tutorial series by Squareline Studio starting here [basics 1](https://www.youtube.com/watch?v=9qp3Lmc8r-Q)
followed by
2: this tutorial which shows how to tie the UI to real world interfaces (like what I did with the RGB LED) https://www.youtube.com/watch?v=PTKjjDPNIdM

Here is the template: **UPDATED 23 September 2024**
[CYD_Squareline_Studio_VSC_Platformio_demo_Template.zip](https://github.com/user-attachments/files/17098521/CYD_Squareline_Studio_VSC_Platformio_demo_Template.zip)

**Comment by thebrando** [comment]
Thanks for this, I was struggling getting my first app running on my cyd. This made it very easy for me to get it going.

**Comment by peteDDD** [comment]
@thebrando   glad to help.  pay it forward when you can.

**Comment by flixmyswitch** [comment]
July 2024 I am having complie issue with this can someone confirm its still good.

**Reply by peteDDD** [reply]
It would be most helpful if you provided more details such as the error messages you are getting, but please nonetheless, try this...

In platformio.ini change
`platform = espressif32`
to
`platform = espressif32@6.5.0`

This will "lock-in" the version of the platform.  I have found lots of issues lately resulting from changes in the platform because the syntax of many commands have changed as a result of recent updates.  

I just compiled the code with this change without issue.

**Reply by flixmyswitch** [reply]
Thanks i know its user error.  I unzip the code into a folder.  Open that folder in VSC and compile it.  I change what you said above.  

I have reread the TXT file instruction and cant see what I am doing wrong

Compiling .pio\build\cyd\src\src\ui\ui.c.o
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/ui/components/../ui.h:15,
                 from src/ui/components/ui_comp_hook.c:6:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/ui/components/../ui.h:15,
                 from src/ui/components/ui_comp.c:6:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pio\build\cyd\src\src\ui\components\ui_comp_hook.c.o] Error 1
*** [.pio\build\cyd\src\src\ui\components\ui_comp.c.o] Error 1
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/ui/components/../ui.h:15,
                 from src/ui/components/ui_comp_screen1.c:6:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/ui/screens/../ui.h:15,
                 from src/ui/screens/ui_Screen1.c:6:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/ui/components/../ui.h:15,
                 from src/ui/components/ui_comp_screen_1.c:6:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pio\build\cyd\src\src\ui\components\ui_comp_screen1.c.o] Error 1
*** [.pio\build\cyd\src\src\ui\screens\ui_Screen1.c.o] Error 1
*** [.pio\build\cyd\src\src\ui\components\ui_comp_screen_1.c.o] Error 1
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/ui/ui.h:15,
                 from src/ui/ui.c:6:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pio\build\cyd\src\src\ui\ui.c.o] Error 1
In file included from .pio/libdeps/cyd/lvgl/src/misc/lv_log.h:16,
                 from .pio/libdeps/cyd/lvgl/lvgl.h:25,
                 from src/main.cpp:17:
.pio/libdeps/cyd/lvgl/src/misc/../lv_conf_internal.h:41:18: fatal error: ../../lv_conf.h: No such file or directory
         #include "../../lv_conf.h"                /*Else assume lv_conf.h is next to the lvgl folder*/
                  ^~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pio\build\cyd\src\src\main.cpp.o] Error 1

**Reply by peteDDD** [reply]
Please read and act on the file "Squareline Studio us projects instructions.txt in the NECESSARY TEMPATE FILES directory.

Note particularly the following: 
    Next you have to copy files to different locations.  These files are in the directory "NECESARY TEMPLATE FILES"
    copy lv_conf.h to .pio\libdeps\cyd\lv_conf.h
    copy User_Setup.h to .pio\libdeps\cyd\TFT_eSPI
    (these paths are noted in the comment at the top of each file)

**Comment by amirzohaib** [comment]
Thanks for the Template, i got it to work. 
I rewrote the Readme (That text file should be at the Parent directory with large "README FIRST!!" written.
I made it a bit more beginner friendly 

TEMPLATE FOR USING CHEAP YELLOW DISPLAY WITH SQUARELINE STUDIO AND PLATFORMIO ARDUINO FRAMEWORK

Extract:
Extract the zip to a folder where you wish to keep your new project.
(You may want to keep a folder just for the template and duplicate that for each new project).
In Visual Studio Code with Platformio, open the template folder with 'Open Project'.
When you first open the folder containing this template, platformio will ask to build the project,
click yes and it will download the necessary libraries and dependencies.
you might need to install a GIT downloader if download fails


Copy lv_conf.h and replace User_Setup.h:
These files have been modified to work with CYD,
These files are in the directory "NECESARY TEMPLATE FILES"
copy lv_conf.h to .pio\libdeps\cyd\lv_conf.h
copy and replace User_Setup.h to .pio\libdeps\cyd\TFT_eSPI
(these paths are noted in the comment at the top of each file)


Test the Template:
This template includes a simple project from Squareline Studio ("SLS")
The project has a blue button.  When it is pressed, the ball will drop and bounce off the button.
It is already in the src/ui directory. You can build and upload this project to your CYD as a test if you like.
Click Platform IO - Project Tasks - cyd - General -  Upload to build and upload the san
Check that all of the fonts you used in the SLS project are turned on.  These start at line 366 of lv_conf.h
"1" indicates that you need the font, "0" indicates that you do not need the font.
For this example you will need #define LV_FONT_MONTSERRAT_16 1 in your lv_conf

CYD2USB Variant:
For Much better colors on the CYD2USB Variant, add all this to the void setup() in main.cpp

tft.setRotation(1); // Landscape orientation  1 =  CYC usb on right, 2 for ver
tft.invertDisplay(1); // if your colors are inverted
tft.writecommand(ILI9341_GAMMASET); //Gamma curve selected
tft.writedata(2);
delay(120);
tft.writecommand(ILI9341_GAMMASET); //Gamma curve selected
tft.writedata(1);

Create and Upload your own SLS Project:
Make sure your project has the following Settings
Width:320
Height: 240
Depth: 16-bit
Board Group: Arduino
Board: Arduino with TFT_eSPI (Note the LVGL Version, make sure its the same as in platfomio.ini lib-dep)
Set UI Files Export Path: the path to your projects's src/ui folder.
for example: C:\Users\owner\Documents\PlatformIO\Projects\CYD Project\src\ui\
Set LVGL Include Path to lvgl.h
From the Template folder, go to \src\ui and delete everything
You will be exporting your SLS ui into that directory and replacing the ball example.
You should now be able to build your project and further edit the main.cpp code to add necessary real-world
interfaces, etc.

**Reply by peteDDD** [reply]
Thanks again for this.  I *finally* updated the zip package and used your Readme to create a new README FIRST.md including a graphic on the Square Line Studio settings.  

THE NEW ZIP PACKAGE IS NOW IN THE FIRST POSTING OF THIS MESSAGE STRING.

**Comment by peteDDD** [comment]
Thanks.  I'm traveling now but will review it when I return.

**Comment by GthiN89** [comment]
Thanks man, extremely helpfull.

**Comment by gpena208777** [comment]
Does this work for the Capacitive USBC variant?

**Comment by peteDDD** [comment]
I don't know what " Does this work for the Capacitive USBC variant" is.
There was only one version of the CYD when I wrote this.

**Comment by peteDDD** [comment]
NOTE THAT THE ZIP HAS BEEN UPDATED TO INCLUDE A FULL README FIRST.md FILE.

See the updated link to the zip in the first message of this string.

**Comment by Roger-505** [comment]
Thank you so much!!

**Comment by smileymiley** [comment]
Hi, demo compiles but won't upload:

A fatal error occurred: This chip is ESP32-S3 not ESP32. Wrong --chip argument?
*** [upload] Error 2

board is ESP32-8048S043
Any suggestions?
Thanks

**Comment by peteDDD** [comment]
Are you using the "Cheap Yellow Display"?  That is not an S3...

**Comment by smileymiley** [comment]
It's yellow and cheap!
Sunton ESP32-8048S043
https://www.makerfabs.com/sunton-esp32-s3-4-3-inch-ips-with-touch.html

Thanks
m

**Comment by peteDDD** [comment]
So our most bananas but some sell for $6.2M.  The CYD board that I was
using is the ESP32-2432S032 which uses a ESP32-WROOM processor.

I like the board you found.  Costs about twice as much but much more
capability that the original CYD.
If the problem is in the upload, not the compile, that is curious.
I am sorry that I can't test it myself since I don't have one... but put
one in my basket for the future.
Looking at the downloads listed on the sales page (links below), they show
the uploading using the Espressif "Flash Download Tool".

Here is a link that should help you a lot:
https://forum.squareline.io/t/squareline-gui-sunton-esp32-8048s043-4-3-800x480-capacitive-touch-display/2130/4
There is a full platformio project there that the fellow created after he
sorted this out.

Please let me know how you make out and what you learn.



download link ESP32-8048S043：
http://pan.jczn1688.com/directlink/1/ESP32%20module/4.3inch_ESP32-8048S043.zip
<http://pan.jczn1688.com/directlink/1/ESP32%20module/4.3inch_ESP32-8048S043.zip?spm=a2g0o.detail.1000023.1.5e05wfeowfeooY&file=4.3inch_ESP32-8048S043.zip>

download link ESP32-4827S043：
http://pan.jczn1688.com/directlink/1/ESP32%20module/4.3inch_ESP32-4827S043.zip
<http://pan.jczn1688.com/directlink/1/ESP32%20module/4.3inch_ESP32-4827S043.zip?spm=a2g0o.detail.1000023.2.5e05wfeowfeooY&file=4.3inch_ESP32-4827S043.zip>

**Comment by ksjkm** [comment]
Thank you so much....I have wasted hours trying to create a template like this.  Very much appreciated.

**Comment by peteDDD** [comment]
I'm glad it helped.  So many have helped me along the way too.  Pay it
forward.

**Comment by joetemus** [comment]
Finally. Couldn't get it to work for ever. Thanks for the help. Worked great.

**Comment by jruizcoll** [comment]
Great! It runs perfectly on ESP32-2432S028

**Comment by JYI-Code** [comment]
Thank you! Thanks to your help, I successfully updated and got my CYD_2432S028R V3 working.
However, when I upload the graphics created in SquareLine Studio, the colors appear completely different on the display. Is there a solution for this, or does this display only support a limited color palette?
![Import](https://github.com/user-attachments/assets/ce03127d-af07-418f-b328-cb2e5cc343e7)

**Reply by amirzohaib** [reply]
the colors are inverted put a
**`tft.invertDisplay(1);`**

**Comment by m-hufschmid** [comment]
Thank you for this very helpful template. It works like a charm and would have saved me a lot of time if I had discovered it earlier.

**Reply by peteDDD** [reply]
I'm still working on the open source for my time machine...

**Comment by peteDDD** [comment]
Go to Chatgpt.com   and type in the following:

In VSCode with platformio, my code compiles but does not link.  I get these
errors:
/Users/len/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld:
.pio/build/ESP32-2432S028R/firmware.elf section .dram0.bss' will not fit in
region dram0_0_seg'
/Users/len/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld:
DRAM segment data does not fit.
/Users/len/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld:
DRAM segment data does not fit.
/Users/len/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld:
region `dram0_0_seg' overflowed by 45936 bytes


You will get a complete explanation.  You probably don't have the ram
settings correct... but read through the details....
